cpp/src/join/quadtree_point_to_nearest_polyline.cu

```diff
@@ -145,74 +157,156 @@ struct compute_quadtree_point_to_nearest_polyline {
-                 quad_idx.end<uint32_t>(),
-                 d_quad_idx.begin());
-
-    // sort (d_poly_idx, d_quad_idx) using d_quad_idx as key => (quad_idxs, poly_idxs)
-    thrust::sort_by_key(
-      rmm::exec_policy(stream), d_quad_idx.begin(), d_quad_idx.end(), d_poly_idx.begin());
-
-    // reduce_by_key using d_quad_idx as the key
-    // exclusive_scan on numbers of polygons associated with a quadrant to create d_poly_idx_offsets
-
-    rmm::device_uvector<uint32_t> d_poly_idx_offsets(num_pairs, stream);
-
-    uint32_t num_quads =
-      thrust::distance(d_poly_idx_offsets.begin(),
-                       thrust::reduce_by_key(rmm::exec_policy(stream),
-                                             d_quad_idx.begin(),
-                                             d_quad_idx.end(),
-                                             thrust::constant_iterator<uint32_t>(1),
-                                             d_quad_idx.begin(),
-                                             d_poly_idx_offsets.begin())
-                         .second);
-
-    d_quad_idx.resize(num_quads, stream);
-    d_poly_idx_offsets.resize(num_quads, stream);
-
-    thrust::exclusive_scan(rmm::exec_policy(stream),
-                           d_poly_idx_offsets.begin(),
-                           d_poly_idx_offsets.end(),
-                           d_poly_idx_offsets.begin());
-
+    auto num_poly_quad_pairs = poly_quad_pairs.num_rows();
+
+    rmm::device_uvector<uint32_t> quad_poly_indices(num_poly_quad_pairs, stream);
+    rmm::device_uvector<uint32_t> quad_point_offsets(num_poly_quad_pairs, stream);
+    rmm::device_uvector<uint32_t> quad_point_lengths(num_poly_quad_pairs, stream);
+
+    auto quad_poly_indices_and_quad_offsets_lengths = make_zip_iterator(
+      // poly_indices
+      poly_quad_pairs.column(0).begin<uint32_t>(),
+      // quad_offsets
+      thrust::make_permutation_iterator(quadtree.column(4).begin<uint32_t>(),
+                                        poly_quad_pairs.column(1).begin<uint32_t>()),
+      // quad_lengths
+      thrust::make_permutation_iterator(quadtree.column(3).begin<uint32_t>(),
+                                        poly_quad_pairs.column(1).begin<uint32_t>()));
+
+    // Copy the quad/polyline indices into local vectors instead of sorting the input columns
+    thrust::copy(
+      rmm::exec_policy(stream),
+      quad_poly_indices_and_quad_offsets_lengths,
+      quad_poly_indices_and_quad_offsets_lengths + num_poly_quad_pairs,
+      make_zip_iterator(
+        quad_poly_indices.begin(), quad_point_offsets.begin(), quad_point_lengths.begin()));
+
+    // Sort the quad/polyline index pairs by quadrant so that the `point_indices` are
+    // produced/reduced in ascending order below.
+    thrust::stable_sort_by_key(
+      rmm::exec_policy(stream),
+      quad_point_offsets.begin(),
+      quad_point_offsets.end(),
+      make_zip_iterator(quad_point_lengths.begin(), quad_poly_indices.begin()));
```

Comment:
These intermediate allocations + sorting could be avoided if the output of `cuspatial::join_quadtree_and_bounding_boxes` were sorted by the `quad_offset` column.